### PR TITLE
feat: Expose tick functions

### PR DESCRIPTION
**Pull request by bjchambers**
This introduces methods to `Timestream` to create tick-streams, and allows passing a function to construct such a tickstream as the predicate for since and sliding windows.

### CODE REVIEW

File: crates/sparrow-compiler/src/ast_to_dfg.rs [hunk 1]
```diff
@@ -242,20 +243,7 @@ pub fn add_to_dfg(
...
-                        tick_node,
-                        tick_node,
-                        FenlType::Concrete(DataType::Boolean),
-                        agg_input.grouping(),
-                        agg_input.time_domain().clone(),
-                        function_name.location().clone(),
-                        None,
-                    ));
-                    return Ok(tick_node);
+                    return add_tick(dfg, behavior, agg_input.value());
```
Comment (jordanrfrazier) [inline crates/sparrow-compiler/src/ast_to_dfg.rs:246]: Ah okay -- so this is still valid for coming from fenl, doing the weird condition input binding, but through python we pass the input directly. Nice.

**Comment by jordanrfrazier** [comment]
This looks like a good pattern

**Comment by bjchambers** [comment]
@jordanrfrazier 

> This looks like a good pattern

Yeah -- I think this makes sense for the Rust side. As noted in #697, we can discuss encapsulating the ticks themselves in the windows:

```
e.sum(window = kd.windows.hourly())
e.sum(window = kd.windows.Session(3, kd.windows.hourly()))
```

as a way of reducing the number of concepts (no such thing as ticks are exposed) and a way to let us experiment with richer windows. For instance, I *think* we can define both `since hourly` and `outputting once at the end of the hour` nearly in Python (using some of the shift-by-1-nanosecond style tricks), and it would let us experiment with the way those concepts fit together over time.

File: python/pysrc/kaskada/_timestream.py [hunk 2]
```diff
@@ -1463,9 +1463,15 @@ def _aggregation(
...
+        predicate = window.predicate
+        if callable(predicate):
+            predicate = predicate(input)
+        return Timestream._call(op, input, *args, predicate, None)
     elif isinstance(window, kd.windows.Sliding):
-        return Timestream._call(op, input, *args, window.predicate, window.duration)
+        predicate = window.predicate
+        if callable(predicate):
+            predicate = predicate(input)
+        return Timestream._call(op, input, *args, predicate, window.duration)
```
Comment (jordanrfrazier) [inline python/pysrc/kaskada/_timestream.py:1474]: We'll probably have to do something similar to this if we want to support ticks in other functions like `filter/if`. I could work on that as a follow-up to this

**Review by jordanrfrazier — APPROVED** [review]
Looks good, the lambda is really clever. I still think we'll want to go the route of offering explicit parameters to the windows themselves, like `tumbling(units, duration, offset)`, which I suppose we could do in a builder pattern 
```
kd.window.Tumbling().duration(5).units(kd.units.Second)
```

**Comment by bjchambers** [comment]
> Looks good, the lambda is really clever. I still think we'll want to go the route of offering explicit parameters to the windows themselves, like `tumbling(units, duration, offset)`, which I suppose we could do in a builder pattern
> 
> ```
> kd.window.Tumbling().duration(5).units(kd.units.Second)
> ```

I think if we go that route, I'd be more inclined to do `kd.window.Tumbling(seconds = 5)` or taking a timedelta, etc. Specifically, we shouldn't define units ourselves.

**Pull request merged**